source/ceylon/logging/LoggerFactory.ceylon

```diff
@@ -0,0 +1,90 @@
+import ceylon.language.meta.model {
+    ClassOrInterface
+}
+import ceylon.collection {
+    MutableMap,
+    HashMap
+}
+
+"`LoggerFactory` can be satisfied to define a custom
+ logging strategy. Typically, an instance of `LoggerFactory`
+ will be be assigned to [[loggerFactory]]."
+shared
+interface LoggerFactory {
+
+    "Function to obtain a [[Logger]] for the given
+     [[Category]].
+
+     Callers intending to wrap the returned logger should
+     provide the [[ClassOrInterface]] of the wrapper.
+     For example:
+
+            FormattingLogger formattingLogger(Category category)
+                =>  FormattingLogger(loggerFactory.logger(
+                        category, `FormattingLogger`));"
+    shared formal
+    Logger logger<Wrapper>(
+            Category category,
+            "The class that will be used to wrap the returned
+             value, if any. This may be used by loggers that
+             log source file and line number information."
+            ClassOrInterface<Wrapper>? wrapper=null)
+            given Wrapper satisfies Object;
+}
```

Comment:
> Yeah but I don't think you need that 

Do you mean https://github.com/ceylon/ceylon-sdk/issues/368 ? Somehow I need a fqcn to pass to interfaces like <http://www.slf4j.org/apidocs/org/slf4j/spi/LocationAwareLogger.html>.

> I can't quite visualize how you propose for that to work

Something like this, but much better, perhaps with `Obtainable`s:

```ceylon

shared alias Mdc => MutableMap<String, String>;
shared Mdc mdc => loggerFactory.mdc;

shared interface LoggerFactory {
    shared formal Logger logger<Wrapper>(...);
    shared formal Mdc mdc;
}
```

The point being that the logger factory method & mdc would be swapped out as an atomic unit.